httpMiddleware.go

```diff
@@ -0,0 +1,199 @@
+type LoggerFunc func(msg string, m map[string]interface{})
+
+func addContextMiddleware(logger zerolog.Logger, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		//eventType := r.Header.Get("X-GitHub-Event")
+		//deliveryID := r.Header.Get("X-GitHub-Delivery")
+
+		ctx := logger.With()
+		//ctx = ctx.Str("github_event_type", eventType)
+		//ctx = ctx.Str("github_delivery_id", deliveryID)
+
+		l := ctx.Logger()
+
+		r = r.WithContext(l.WithContext(r.Context()))
+
+		h.ServeHTTP(w, r)
+	})
+}
+
+func loggerMiddleware(logger LoggerFunc, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		statusWriter := &statusCodeResponseWriter{w, http.StatusOK}
+		start := time.Now()
+		h.ServeHTTP(statusWriter, r)
+
+		// request duration in miliseconds
+		duration := time.Since(start).Nanoseconds() / 1e6
+		statusCode := statusWriter.statusCode
+
+		logMessage := fmt.Sprintf("[%d] %s %s", statusCode, r.Method, r.URL.String())
+		logMap := map[string]interface{}{
+			"responseTime": fmt.Sprintf("%d", duration),
+		}
+
+		logger(logMessage, logMap)
+	})
+}
+
+func inboundMetricsMiddleware(promRegisterer prometheus.Registerer, h http.Handler) http.Handler {
+
+	httpRequests := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "inbound_http_status_code_total",
+			Help: "Counter of HTTP status codes of inbound HTTP requests",
+		},
+		[]string{"status_code", "method"},
+	)
+	promRegisterer.MustRegister(httpRequests)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		statusWriter := &statusCodeResponseWriter{w, http.StatusOK}
+		//start := time.Now()
+		h.ServeHTTP(statusWriter, r)
+
+		// request duration in miliseconds
+		//duration := time.Since(start).Nanoseconds() / 1e6
+		statusCode := statusWriter.statusCode
+
+		httpRequests.WithLabelValues(strconv.Itoa(statusCode), r.Method).Inc()
+
+	})
+}
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+type ClientMiddleware func(http.RoundTripper) http.RoundTripper
+
+func (fn roundTripperFunc) RoundTrip(r *http.Request) (*http.Response, error) {
+	return fn(r)
+}
+
+func clientMetricsMiddleware(promRegisterer prometheus.Registerer, requestDestination string) func(http.RoundTripper) http.RoundTripper {
+
+	metricHTTPRequestsStatus := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "outbound_http_status_code_total",
+			Help:        "Counter of HTTP status codes of outbound HTTP requests",
+			ConstLabels: prometheus.Labels{"destination": requestDestination},
+		},
+		[]string{"status_code"},
+	)
+
+	promRegisterer.MustRegister(metricHTTPRequestsStatus)
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+
+			res, err := next.RoundTrip(r)
+			if err != nil {
+				return res, err
+			}
+			if res == nil {
+				return res, err
+			}
```

Comment:
Do you know when this can happen? ie. `err` AND `res` are `nil` from a RoundTrip method?

Reply:
No, I don't know when this can happen. I was inspired by how palentir approached this in go-githubapp https://github.com/palantir/go-githubapp/blob/8e893421b1e560ebc905cf48adbd5c1d09f0e899/githubapp/middleware.go#L66

But here they don't look at `err`

Reply:
https://github.com/lunarway/release-manager-bot/pull/21/commits/10eebfa9ec8b13c92c28d45a8f062cb080173e80